### ISSUE TITLE
[Improvement] Better support for  AQE sortMergeJoin -> broadcastHashJoin

### ISSUE THREAD

**Issue by zuston** [apache/incubator-uniffle#838]
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-uniffle/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

In #630 , we disable the `localShuffleRead` for AQE in spark. For a pity, sometimes the partial partition is so huge, but the broadcastHashJoin won't optimize skew join. This optimization will be solved by spark community ASSP.

After that, I will propose a new patch for uniffle to better support Spark older version.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

**Comment by zuston** [comment]
cc @advancedxy @jerqi

**Comment by jerqi** [comment]
I can't get your point. If you want to push a Spark's feature patch to `patch` directory of project Uniffle, I will vote -1.

**Comment by zuston** [comment]
> This optimization will be solved by spark community ASSP. After that, I will propose a new patch for uniffle to better support Spark older version.

I mean this optimization should be proposed to Spark community. After that, I will propose a new patch to older spark version to support this optimization.

**Comment by jerqi** [comment]
> > This optimization will be solved by spark community ASSP. After that, I will propose a new patch for uniffle to better support Spark older version.
> 
> I mean this optimization should be proposed to Spark community. After that, I will propose a new patch to older spark version to support this optimization.

How do you support the old spark versions?

**Comment by zuston** [comment]
like spark3.1.1

**Comment by jerqi** [comment]
> like spark3.1.1

I wanna ask you how not what.

**Comment by zuston** [comment]
https://github.com/apache/incubator-uniffle/tree/master/patch/spark. Like this.

I don't quite understand what is unclear about my expression.

**Comment by jerqi** [comment]
> https://github.com/apache/incubator-uniffle/tree/master/patch/spark. Like this.
> 
> I don't quite understand what is unclear about my expression.

I reject. Because we should avoid modifying the Spark code. It's unrelated to Uniffle although it will improve the performance of Uniffle. If you need this, you should cherry-pick this feature to your internal branch.

**Comment by zuston** [comment]
> I reject. Because we should avoid modifying the Spark code. It's unrelated to Uniffle although it will improve the performance of Uniffle. If you need this, you should cherry-pick this feature to your internal branch.

It's OK.

The reason of proposing this is to expose the problem of using uniffle(RSS) with spark AQE poor performance. Those using uniffle should pay more attension on this. And all optimization about this could be tracked here. 

If you reject this, we could keep this issue open and attach the related spark optimization PR here if exists.

**Comment by jerqi** [comment]
You can modify the document https://github.com/apache/incubator-uniffle-website/blob/master/docs/07-tuning.md.

**Comment by zuston** [comment]
Let's discuss first if anyone is interested

**Comment by pan3793** [comment]
Kyuubi project tackles this issue by allowing to forcibly shuffle before the join operator https://github.com/apache/kyuubi/blob/master/extensions/spark/kyuubi-extension-spark-3-3/src/main/scala/org/apache/kyuubi/sql/InsertShuffleNodeBeforeJoin.scala#L38

**Comment by zuston** [comment]
Oh, it has been proposed but still reviewing. https://github.com/apache/spark/pull/40312

**Comment by zuston** [comment]
> In my opinion, this is just another optimize SkewJoin rule, which should be integrated into your internal Spark code if it's not merged into Spark upstream yet. Optimizing skew joins is out of the scope of Uniffle project. However, if these optimization rules is well maintained elsewhere out of Uniffle project(maybe kyubi?), I believe the community is more than happy to add it to the tuning document in details.

Yes. I have to re-claim the thought, just to share some performance ideas for Spark & Uniffle. I’m OK for keeping this patch out of uniffle.

**Comment by zuston** [comment]
> Kyuubi project tackles this issue by allowing to forcibly shuffle before the join operator https://github.com/apache/kyuubi/blob/master/extensions/spark/kyuubi-extension-spark-3-3/src/main/scala/org/apache/kyuubi/sql/InsertShuffleNodeBeforeJoin.scala#L38

Good way!
